relayer/relayer.go

```diff
@@ -0,0 +1,157 @@
+		} else {
+			rpcEndpoint = "http://localhost:8545"
+		}
+	}
+
+	return rpcEndpoint
+}
+
+func NewRelayer(
+	dataDir string,
+	rpcEndpoint string,
+	stateReceiverAddr ethgo.Address,
+	logger hcf.Logger,
+	key ethgo.Key,
+) *Relayer {
+	endpoint := sanitizeRPCEndpoint(rpcEndpoint)
+
+	// create the JSON RPC client
+	client, err := jsonrpc.NewClient(endpoint)
+	if err != nil {
+		logger.Error("Failed to create the JSON RPC client", "err", err)
+
+		return nil
+	}
+
+	txRelayer, err := txrelayer.NewTxRelayer(txrelayer.WithClient(client))
+
+	return &Relayer{
+		dataDir:           dataDir,
+		rpcEndpoint:       endpoint,
+		stateReceiverAddr: stateReceiverAddr,
+		logger:            logger,
+		client:            client,
+		txRelayer:         txRelayer,
+		key:               key,
+	}
+}
+
+func (r *Relayer) Start() error {
+	et := eventTracker{
+		dataDir:           r.dataDir,
+		rpcEndpoint:       r.rpcEndpoint,
+		stateReceiverAddr: r.stateReceiverAddr,
+		subscriber:        r,
+		logger:            r.logger,
+	}
+
+	return et.start()
+}
+
+func (r *Relayer) AddLog(log *ethgo.Log) {
+	r.logger.Info("Received a log", "log", log)
+
+	if commitEvent.Match(log) {
+		vals, err := commitEvent.ParseLog(log)
+		if err != nil {
+			panic(err)
+		}
+
+		var startID uint64
+		if sid, ok := vals["startId"].(big.Int); ok {
+			startID = sid.Uint64()
+		}
+
+		var endID uint64
+		if eid, ok := vals["endId"].(big.Int); ok {
+			endID = eid.Uint64()
+		}
+
+		r.logger.Info("Commit: Block %d StartID %d EndID %d\n", log.BlockNumber, startID, endID)
+
+		for i := startID; i < endID; i++ {
+			// query the state sync proof
+			stateSyncProof, err := r.queryStateSyncProof(strconv.Itoa(int(i)))
+			if err != nil {
+				r.logger.Error("Failed to query state sync proof", "err", err)
+			}
+
+			if err := r.executeStateSync(stateSyncProof); err != nil {
+				r.logger.Error("Failed to execute state sync", "err", err)
+			}
+		}
+	}
+}
+
+// queryStateSyncProof queries the state sync proof
+func (r *Relayer) queryStateSyncProof(stateSyncID string) (*types.StateSyncProof, error) {
+	// retrieve state sync proof
+	var stateSyncProof types.StateSyncProof
+
+	err := r.client.Call("bridge_getStateSyncProof", &stateSyncProof, stateSyncID)
+	if err != nil {
+		return nil, err
+	}
+
+	r.logger.Debug("state sync proof:", stateSyncProof)
+
+	return &stateSyncProof, nil
+}
+
+// executeStateSync executes the state sync
+func (r *Relayer) executeStateSync(stateSyncProof *types.StateSyncProof) error {
+	input, err := types.ExecuteStateSyncABIMethod.Encode(
+		[2]interface{}{stateSyncProof.Proof, stateSyncProof.StateSync.ToMap()},
+	)
+	if err != nil {
+		return err
+	}
+
+	// execute the state sync
+	txn := &ethgo.Transaction{
+		From:     r.key.Address(),
+		To:       (*ethgo.Address)(&contracts.StateReceiverContract),
+		GasPrice: 0,
+		Gas:      types.StateTransactionGasLimit,
+		Input:    input,
+	}
+
+	_, err = r.txRelayer.SendTransaction(txn, r.key)
```

Comment:
Would you like to check receipt status and return an error if its status is failed?

Reply:
Nice one, done in https://github.com/0xPolygon/polygon-edge/pull/1044/commits/d083e8096a78e2777d4da633efee1d706d273973